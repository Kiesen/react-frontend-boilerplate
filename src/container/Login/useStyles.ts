import { makeStyles } from '@material-ui/core/styles';

/**
 * Mui styling for the login container component.
 *
 * @exports StylesHook
 */
export default makeStyles({
  root: {
    minHeight: '500px',
    height: '100vh',
    display: 'flex',
    justifyContent: 'center',
    flexDirection: 'column',
  },
});
