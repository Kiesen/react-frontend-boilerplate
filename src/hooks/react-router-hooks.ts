/**
 * All hooks are related to issue 6430 of the react router repository:
 * https://github.com/ReactTraining/react-router/issues/6430
 */
import { useContext } from 'react';
import { __RouterContext as RouterContext } from 'react-router';

/**
 * Custom hook that returns the react router object.
 * Uses the react context hook with the react router
 * context as an argument.
 *
 * @return {Object} router - router object
 */
export const useRouter = () => {
  return useContext(RouterContext);
};

/**
 * Custom hook that just returns the
 * react router params.
 *
 * @return {Object} params - param object
 */
export const useParams = () => {
  const { match } = useRouter();
  return match.params;
};

/**
 * Custom hook that returns the router location object
 * and a custom function that uses push or replace
 * to change the route.
 *
 * @return an object containing location and navigate.
 */
export const useLocation = () => {
  const { location, history } = useRouter();

  /**
   * Custom function that uses the react router history to
   * either push the new route on the history stack
   * or replace the current route.
   *
   * @param {string} to - the new location
   * @param {boolean} [replace] - should replace be used
   *
   */
  const navigate = (to: string, replace = false) => {
    if (replace) {
      history.replace(to);
    } else {
      history.push(to);
    }
  };

  return {
    location,
    navigate,
  };
};
