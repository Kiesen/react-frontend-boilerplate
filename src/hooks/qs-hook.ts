import qs from 'qs';

import { useRouter } from 'hooks/react-router-hooks';

interface Props {
  [key: string]: string | null;
}
type SetQueryString = () => (props: Props) => void;

/**
 * Custom hook that returns a callback.
 * The callback function takes a one dimensional object as an argument,
 * creates a new query string and pushes the query string to the
 * history stack.
 *
 * Important note: If you want to remove any keys from the query string
 * just provide the key with a null value e.g. { x: null }
 *
 * @return callback with following signature (queryObject) => void
 */
export const useSetQueryString: SetQueryString = () => {
  const { location, history } = useRouter();

  return (props: Props) => {
    const queryString = qs.stringify(
      {
        ...qs.parse(location.search.slice(1)),
        ...qs.parse(qs.stringify(props, { strictNullHandling: true }), {
          strictNullHandling: true,
        }),
      },
      { addQueryPrefix: true, skipNulls: true }
    );
    history.push(`${location.pathname}${queryString}${location.hash}`);
  };
};
