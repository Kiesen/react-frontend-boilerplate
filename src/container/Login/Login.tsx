import React, {
  FC,
  useState,
  SyntheticEvent,
  ChangeEvent,
  FocusEvent,
  KeyboardEvent,
} from 'react';
import { useTranslation } from 'react-i18next';

import CardContent from '@material-ui/core/CardContent';
import Card from '@material-ui/core/Card';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Grid from '@material-ui/core/Grid';

import useStyles from 'container/Login/useStyles';
import Footer from 'container/Footer/Footer';
import { useLocation } from 'hooks/react-router-hooks';
import { AUTH_CREDENTIALS } from 'constants/storageKeys';
import { BASE } from 'constants/routes';

/**
 * Basic login container component.
 * Handels onChange events of the email and password input fields
 * and checks if the values are valid after a login action got triggered
 * or an input field triggers the onBlur event.
 *
 * @returns Login page with input fields
 */
const Login: FC<{}> = () => {
  // CSS classes created by material ui
  const classes = useStyles();
  // React router context
  const { navigate } = useLocation();
  // i18n translations
  const { t } = useTranslation();
  // State used for the text fields
  const [values, setValues] = useState({
    email: '',
    password: '',
    emailError: false,
    passwordError: false,
  });

  // Mail regex taken from the w3 recommendation:
  // https://www.w3.org/TR/html5/forms.html#valid-e-mail-address
  const emailRe = new RegExp(
    "^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9]" +
      '(?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:.[a-zA-Z0-9]' +
      '(?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+$'
  );

  /**
   * Function triggers all necessary validation steps and
   * pushes the BASE path to the history stack, if no error occures.
   *
   * @param {SyntheticEvent} event - SyntheticEvent<HTMLButtonElement>
   */
  const loginAction = (event?: SyntheticEvent<HTMLButtonElement>): void => {
    if (event) event.preventDefault();
    const emailError = !emailRe.test(values.email);
    const passwordError = values.password.length < 1;
    if (emailError || passwordError) {
      setValues({
        ...values,
        emailError: emailError,
        passwordError: passwordError,
      });
    } else {
      // -- TODO --
      // Add login call to your api or auth service here
      // Also add some error handling if auth failes
      sessionStorage.setItem(AUTH_CREDENTIALS, '');
      navigate(BASE);
    }
  };

  /**
   * Function set's the input field value to the component state.
   *
   * @param {ChangeEvent} event - ChangeEvent<HTMLInputElement>
   */
  const handleChange = (event: ChangeEvent<HTMLInputElement>): void => {
    event.preventDefault();
    setValues({ ...values, [event.target.name]: event.target.value });
  };

  /**
   * Function triggers the loginAction if an enter key got pressed.
   *
   * @param {KeyboardEvent} event - KeyboardEvent<HTMLInputElement>
   */
  const handleKeyPress = (event: KeyboardEvent<HTMLInputElement>): void => {
    if (event.key === 'Enter') loginAction();
  };

  /**
   * Function validate a input field if a blur event got triggered.
   *
   * @param {FocusEvent} event - FocusEvent<HTMLInputElement>
   */
  const validateOnBlur = (event: FocusEvent<HTMLInputElement>): void => {
    event.preventDefault();
    switch (event.target.name) {
      case 'email':
        setValues({ ...values, emailError: !emailRe.test(event.target.value) });
        break;
      case 'password':
        setValues({ ...values, passwordError: event.target.value.length < 1 });
        break;
      default:
        break;
    }
  };

  return (
    <Grid className={classes.root}>
      <Grid container alignItems="center" justify="center">
        <Card>
          <CardContent>
            <h1>{t('client:headline.login')}</h1>
          </CardContent>
          <CardContent>
            <form noValidate>
              <TextField
                label={t('client:form.label.email')}
                fullWidth
                type="email"
                name="email"
                margin="normal"
                variant="outlined"
                value={values.email}
                error={values.emailError}
                onChange={handleChange}
                onBlur={validateOnBlur}
                onKeyPress={handleKeyPress}
                helperText={
                  values.emailError ? t('client:form.errorText.email') : ''
                }
              />
              <TextField
                label={t('client:form.label.password')}
                fullWidth
                type="password"
                name="password"
                margin="normal"
                variant="outlined"
                value={values.password}
                error={values.passwordError}
                onChange={handleChange}
                onBlur={validateOnBlur}
                onKeyPress={handleKeyPress}
                helperText={
                  values.passwordError
                    ? t('client:form.errorText.password')
                    : ''
                }
              />
            </form>
          </CardContent>
          <CardContent>
            <Button
              fullWidth
              variant="contained"
              color="primary"
              onClick={loginAction}
            >
              {t('client:button.login')}
            </Button>
          </CardContent>
        </Card>
      </Grid>
      <Footer />
    </Grid>
  );
};

export default Login;
