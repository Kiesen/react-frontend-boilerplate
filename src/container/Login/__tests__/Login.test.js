import React from 'react';
import { shallow } from 'enzyme';
import * as hooks from 'hooks/react-router-hooks';

import Login from 'container/Login/Login';

describe('<Login />', () => {
  beforeEach(() => {
    jest.spyOn(hooks, 'useLocation').mockImplementation(() => ({
      navigate: jest.fn(),
    }));
  });

  it('renders correctly', () => {
    const wrapper = shallow(<Login />);
    expect(wrapper).toMatchSnapshot();
  });
});
