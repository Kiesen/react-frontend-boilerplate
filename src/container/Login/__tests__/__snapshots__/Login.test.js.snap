// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> renders correctly 1`] = `
<WithStyles(ForwardRef(Grid))
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    justify="center"
  >
    <WithStyles(ForwardRef(Card))>
      <WithStyles(ForwardRef(CardContent))>
        <h1>
          client:headline.login
        </h1>
      </WithStyles(ForwardRef(CardContent))>
      <WithStyles(ForwardRef(CardContent))>
        <form
          noValidate={true}
        >
          <WithStyles(ForwardRef(TextField))
            error={false}
            fullWidth={true}
            helperText=""
            label="client:form.label.email"
            margin="normal"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyPress={[Function]}
            type="email"
            value=""
            variant="outlined"
          />
          <WithStyles(ForwardRef(TextField))
            error={false}
            fullWidth={true}
            helperText=""
            label="client:form.label.password"
            margin="normal"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyPress={[Function]}
            type="password"
            value=""
            variant="outlined"
          />
        </form>
      </WithStyles(ForwardRef(CardContent))>
      <WithStyles(ForwardRef(CardContent))>
        <WithStyles(ForwardRef(Button))
          color="primary"
          fullWidth={true}
          onClick={[Function]}
          variant="contained"
        >
          client:button.login
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(CardContent))>
    </WithStyles(ForwardRef(Card))>
  </WithStyles(ForwardRef(Grid))>
  <Footer />
</WithStyles(ForwardRef(Grid))>
`;
